### PR TITLE
[d4hines] Benchmarking: ticket transfers                 d4hines/ticket-transfer-benchmarking

### PR DESCRIPTION
**Pull request by d4hines**
## Depends
- [ ] #647 

## Problem

We want a reproducible way of testing through the throughput of Deku. This will help us determine where our bottlenecks are and where to focus optimization effrots.

## Solution

This PR adds an e2e test similar to that in #470 that attempts to saturate the network with ticket transfer transactions. 

There is a lot that could be improved in this PR:
- I have not tested the efficiency of the load tester. Obviously, if your load tester is not fast enough, you won't fully exercise your system. So I can't say with confidence that this truly saturates the network.
- We obviously need to add more variety to the tests and the testing conditions (different network topologies, killing nodes etc.).
- I'd like to automate the execution and reporting of these benchmarks but I need help from infra to do that (cc @ulrikstrid).

Despite the above, I think this is a good start, and will already provide immediately useful information to us for debugging performance issues (it already helped me detect #538), and I'm interested in feedback to get it merged in. 

### Stacked PR Chain: d4hines
| PR | Title | Status |  Merges Into  |
|:--:|:------|:-------|:-------------:|
|#691|fix: fix prometheus with tilt||-|
|#647|feat: block rate and transaction rate metrics  d4hines/block-rate-and-tps||#691|
|#658|chore: refactor cli terms                      d4hines/batch-operations||#647|
|#681|chore: refactor config out of node state||#658|
|#668|make minimum block time configurable||#658|
|#539|👉Benchmarking: ticket transfers                 d4hines/ticket-transfer-benchmarking||#658|

### CODE REVIEW

File: tests/e2e/load_test.ml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  in
+  Lwt.async (fun () ->
+      Networking.request_user_operation_gossip
+        { user_operation = transaction }
+        validator_uri);
+  transaction
+
+let spam_transactions ~ticketer n () =
+  (* Hard-coded for now. TODO: get these dynamically, see https://github.com/marigold-dev/deku/pull/450 *)
+  let validators_uris =
```
Comment (lykimq) [inline tests/e2e/load_test.ml:72]: You can call the function `validators_uris`, it is the same for `get_random_validator_uri`

Comment (d4hines) [reply tests/e2e/load_test.ml:72]: Addressed :thumbsup:

**Comment by d4hines** [comment]
Closing to consolidate the PR's.

**Comment by d4hines** [comment]
Woops, I mean to keep this PR open

**Review by gsebil08 — APPROVED** [review]
👍

File: src/node/flows.ml [hunk 2]
```diff
@@ -396,19 +396,20 @@ let received_user_operation state update_state user_operation =
   let operation_exists =
     Node.Operation_map.mem operation state.Node.pending_operations in
 
-  if not operation_exists then (
-    Lwt.async (fun () ->
-        broadcast_user_operation_gossip state { user_operation });
-    append_operation state update_state operation);
+  if not operation_exists then
+    (* TODO: this broadcast is unnecessary in many case. We should optimize the messages sent. *)
```
[the remark this answers is not in the document]
Comment (d4hines) [reply src/node/flows.ml]: When a node receives an operation it's never seen before, it broadcasts that message to every other node. This results in quadratic message behavior. Pollinate is "right" solution to this :smile:

File: sandbox.sh [hunk 3]
```diff
@@ -516,6 +523,9 @@ deploy-dummy-ticket)
 deposit-dummy-ticket)
   deposit_ticket
   ;;
+load-test)
```
Comment (d4hines) [inline sandbox.sh:526]: @lykimq we can make a new PR that adds a `--added_latency` option that uses https://www.pico.net/kb/how-can-i-simulate-delayed-and-dropped-packets-in-linux/ to inject extra latency.

**Review by lykimq — APPROVED** [review]

**Pull request merged**